Recipe unit tests make network calls to container registry

## Bug information

Some of the unit tests for recipes seem to be making network calls. We should find a way to mock these outbound calls.

```
=== Failed
=== FAIL: pkg/recipes/driver Test_Bicep_Execute_SimulatedEnvironment (0.39s)
    bicep.go:84: "ts"="2023-10-16 23:57:53.636530" "level"=0 "msg"="Deploying recipe: \"test-recipe\", template: \"ghcr.io/radius-project/dev/recipes/functionaltest/parameters/mongodatabases/azure:1.0\""
    bicep_test.go:382: 
        	Error Trace:	/home/runner/work/radius/radius/pkg/recipes/driver/bicep_test.go:382
        	Error:      	Received unexpected error:
        	            	code RecipeDownloadFailed: err code RecipeLanguageFailure: err failed to fetch repository from the path "ghcr.io/radius-project/dev/recipes/functionaltest/parameters/mongodatabases/azure:1.0": HEAD "https://ghcr.io/v2/radius-project/dev/recipes/functionaltest/parameters/mongodatabases/azure/manifests/1.0": GET "https://ghcr.io/token?scope=repository%3Aradius-project%2Fdev%2Frecipes%2Ffunctionaltest%2Fparameters%2Fmongodatabases%2Fazure%3Apull&service=ghcr.io": response status code 403: denied: requested access to the resource is denied
        	Test:       	Test_Bicep_Execute_SimulatedEnvironment

=== FAIL: pkg/recipes/driver Test_Bicep_GetRecipeMetadata_Success (0.13s)
    bicep_test.go:488: 
        	Error Trace:	/home/runner/work/radius/radius/pkg/recipes/driver/bicep_test.go:488
        	Error:      	Received unexpected error:
        	            	code RecipeLanguageFailure: err failed to fetch repository from the path "ghcr.io/radius-project/dev/recipes/functionaltest/parameters/mongodatabases/azure:1.0": HEAD "https://ghcr.io/v2/radius-project/dev/recipes/functionaltest/parameters/mongodatabases/azure/manifests/1.0": GET "https://ghcr.io/token?scope=repository%3Aradius-project%2Fdev%2Frecipes%2Ffunctionaltest%2Fparameters%2Fmongodatabases%2Fazure%3Apull&service=ghcr.io": response status code 403: denied: requested access to the resource is denied
        	Test:       	Test_Bicep_GetRecipeMetadata_Success

=== FAIL: pkg/recipes/driver Test_Bicep_GetRecipeMetadata_Error (0.13s)
    bicep_test.go:515: 
        	Error Trace:	/home/runner/work/radius/radius/pkg/recipes/driver/bicep_test.go:515
        	Error:      	Not equal: 
        	            	expected: &recipes.RecipeError{ErrorDetails:v1.ErrorDetails{Code:"RecipeLanguageFailure", Message:"failed to fetch repository from the path \"ghcr.io/radius-project/dev/test-non-existent-recipe\": HEAD \"[https://ghcr.io/v2/radius-project/dev/test-non-existent-recipe/manifests/latest\](https://ghcr.io/v2/radius-project/dev/test-non-existent-recipe/manifests/latest/)": GET \"https://ghcr.io/token?scope=repository%3Aradius-project%2Fdev%2Ftest-non-existent-recipe%3Apull&service=ghcr.io\": response status code 403: denied: requested access to the resource is denied", Target:"", AdditionalInfo:[]v1.ErrorAdditionalInfo(nil), Details:[]v1.ErrorDetails(nil)}, DeploymentStatus:"setupError"}
        	            	actual  : &recipes.RecipeError{ErrorDetails:v1.ErrorDetails{Code:"RecipeLanguageFailure", Message:"failed to fetch repository from the path \"ghcr.io/radius-project/dev/test-non-existent-recipe\": ghcr.io/radius-project/dev/test-non-existent-recipe:latest: not found", Target:"", AdditionalInfo:[]v1.ErrorAdditionalInfo(nil), Details:[]v1.ErrorDetails(nil)}, DeploymentStatus:"setupError"}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -3,3 +3,3 @@
        	            	   Code: (string) (len=21) "RecipeLanguageFailure",
        	            	-  Message: (string) (len=379) "failed to fetch repository from the path \"ghcr.io/radius-project/dev/test-non-existent-recipe\": HEAD \"[https://ghcr.io/v2/radius-project/dev/test-non-existent-recipe/manifests/latest\](https://ghcr.io/v2/radius-project/dev/test-non-existent-recipe/manifests/latest/)": GET \"https://ghcr.io/token?scope=repository%3Aradius-project%2Fdev%2Ftest-non-existent-recipe%3Apull&service=ghcr.io\": response status code 403: denied: requested access to the resource is denied",
        	            	+  Message: (string) (len=165) "failed to fetch repository from the path \"ghcr.io/radius-project/dev/test-non-existent-recipe\": ghcr.io/radius-project/dev/test-non-existent-recipe:latest: not found",
        	            	   Target: (string) "",
        	Test:       	Test_Bicep_GetRecipeMetadata_Error

DONE 3018 tests, 9 skipped, 3 failures in 680.692s
make: *** [build/test.mk:47: test] Error 1
Error: Process completed with exit code 2.
```

[AB#9827](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/9827)

--- a/pkg/recipes/driver/bicep.go
+++ b/pkg/recipes/driver/bicep.go
@@ -25,10 +25,12 @@ import (
 	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
 	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
 	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources"
+	"github.com/go-logr/logr"
 	"github.com/radius-project/radius/pkg/to"
 	"golang.org/x/sync/errgroup"
+	"oras.land/oras-go/v2/registry/remote"
 
-	"github.com/go-logr/logr"
+	coredm "github.com/radius-project/radius/pkg/corerp/datamodel"
 	"github.com/radius-project/radius/pkg/metrics"
 	"github.com/radius-project/radius/pkg/portableresources/datamodel"
 	"github.com/radius-project/radius/pkg/portableresources/processors"
@@ -41,8 +43,6 @@ import (
 	"github.com/radius-project/radius/pkg/ucp/resources"
 	resources_radius "github.com/radius-project/radius/pkg/ucp/resources/radius"
 	"github.com/radius-project/radius/pkg/ucp/ucplog"
-
-	coredm "github.com/radius-project/radius/pkg/corerp/datamodel"
 )
 
 //go:generate mockgen -destination=./mock_driver.go -package=driver -self_package github.com/radius-project/radius/pkg/recipes/driver github.com/radius-project/radius/pkg/recipes/driver Driver
@@ -74,6 +74,9 @@ type bicepDriver struct {
 	DeploymentClient *clients.ResourceDeploymentsClient
 	ResourceClient   processors.ResourceClient
 	options          BicepOptions
+
+	// RegistryClient is the optional client used to interact with the container registry.
+	RegistryClient remote.Client
 }
 
 // Execute fetches recipe contents from container registry, creates a deployment ID, a recipe context parameter, recipe parameters,
@@ -85,7 +88,8 @@ func (d *bicepDriver) Execute(ctx context.Context, opts ExecuteOptions) (*recipe
 
 	recipeData := make(map[string]any)
 	downloadStartTime := time.Now()
-	err := util.ReadFromRegistry(ctx, opts.Definition, &recipeData)
+
+	err := util.ReadFromRegistry(ctx, opts.Definition, &recipeData, d.RegistryClient)
 	if err != nil {
 		metrics.DefaultRecipeEngineMetrics.RecordRecipeDownloadDuration(ctx, downloadStartTime,
 			metrics.NewRecipeAttributes(metrics.RecipeEngineOperationDownloadRecipe, opts.Recipe.Name, &opts.Definition, recipes.RecipeDownloadFailed))
@@ -258,7 +262,7 @@ func (d *bicepDriver) GetRecipeMetadata(ctx context.Context, opts BaseOptions) (
 	//		}
 	//	}
 	recipeData := make(map[string]any)
-	err := util.ReadFromRegistry(ctx, opts.Definition, &recipeData)
+	err := util.ReadFromRegistry(ctx, opts.Definition, &recipeData, d.RegistryClient)
 	if err != nil {
 		return nil, err
 	}

--- a/pkg/recipes/driver/bicep.go
+++ b/pkg/recipes/driver/bicep.go
@@ -26,7 +26,6 @@ import (
 	"github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime"
 	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources"
 	"github.com/go-logr/logr"
-	"github.com/radius-project/radius/pkg/to"
 	"golang.org/x/sync/errgroup"
 	"oras.land/oras-go/v2/registry/remote"
 
@@ -39,7 +38,8 @@ import (
 	recipes_util "github.com/radius-project/radius/pkg/recipes/util"
 	"github.com/radius-project/radius/pkg/rp/util"
 	rpv1 "github.com/radius-project/radius/pkg/rp/v1"
-	clients "github.com/radius-project/radius/pkg/sdk/clients"
+	"github.com/radius-project/radius/pkg/sdk/clients"
+	"github.com/radius-project/radius/pkg/to"
 	"github.com/radius-project/radius/pkg/ucp/resources"
 	resources_radius "github.com/radius-project/radius/pkg/ucp/resources/radius"
 	"github.com/radius-project/radius/pkg/ucp/ucplog"

--- a/pkg/recipes/driver/bicep_test.go
+++ b/pkg/recipes/driver/bicep_test.go
@@ -18,21 +18,24 @@ package driver
 
 import (
 	"fmt"
+	"strings"
 	"testing"
 
 	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources"
-	gomock "github.com/golang/mock/gomock"
 	v1 "github.com/radius-project/radius/pkg/armrpc/api/v1"
 	corerp_datamodel "github.com/radius-project/radius/pkg/corerp/datamodel"
 	"github.com/radius-project/radius/pkg/portableresources/processors"
 	"github.com/radius-project/radius/pkg/recipes"
 	"github.com/radius-project/radius/pkg/recipes/recipecontext"
+	"github.com/radius-project/radius/pkg/rp/util/registrytest"
 	rpv1 "github.com/radius-project/radius/pkg/rp/v1"
 	clients "github.com/radius-project/radius/pkg/sdk/clients"
 	"github.com/radius-project/radius/pkg/to"
 	"github.com/radius-project/radius/pkg/ucp/resources"
 	resources_kubernetes "github.com/radius-project/radius/pkg/ucp/resources/kubernetes"
 	"github.com/radius-project/radius/test/testcontext"
+
+	"github.com/golang/mock/gomock"
 	"github.com/stretchr/testify/require"
 )
 
@@ -376,7 +379,9 @@ func Test_Bicep_PrepareRecipeResponse_EmptyResult(t *testing.T) {
 }
 
 func Test_Bicep_Execute_SimulatedEnvironment(t *testing.T) {
-	t.Skip("This test makes outbound calls. #6490")
+	ts := registrytest.NewFakeRegistryServer(t)
+	t.Cleanup(ts.CloseServer)
+
 	opts := ExecuteOptions{
 		BaseOptions: BaseOptions{
 			Configuration: recipes.Configuration{
@@ -395,13 +400,13 @@ func Test_Bicep_Execute_SimulatedEnvironment(t *testing.T) {
 			Definition: recipes.EnvironmentDefinition{
 				Name:         "test-recipe",
 				Driver:       recipes.TemplateKindBicep,
-				TemplatePath: "ghcr.io/radius-project/dev/recipes/functionaltest/parameters/mongodatabases/azure:1.0",
+				TemplatePath: ts.TestImageURL,
 				ResourceType: "Applications.Datastores/mongoDatabases",
 			},
 		},
 	}
 	ctx := testcontext.New(t)
-	d := &bicepDriver{}
+	d := &bicepDriver{RegistryClient: ts.TestServer.Client()}
 	recipesOutput, err := d.Execute(ctx, opts)
 	require.NoError(t, err)
 	require.Nil(t, recipesOutput)
@@ -489,20 +494,21 @@ func Test_Bicep_Delete_Error(t *testing.T) {
 }
 
 func Test_Bicep_GetRecipeMetadata_Success(t *testing.T) {
-	t.Skip("This test makes outbound calls. #6490")
+	ts := registrytest.NewFakeRegistryServer(t)
+	t.Cleanup(ts.CloseServer)
+
 	ctx := testcontext.New(t)
-	driver := bicepDriver{}
+	driver := &bicepDriver{RegistryClient: ts.TestServer.Client()}
 	recipeDefinition := recipes.EnvironmentDefinition{
 		Name:         "mongo-azure",
 		Driver:       recipes.TemplateKindBicep,
-		TemplatePath: "ghcr.io/radius-project/dev/recipes/functionaltest/parameters/mongodatabases/azure:1.0",
+		TemplatePath: ts.TestImageURL,
 		ResourceType: "Applications.Datastores/mongoDatabases",
 	}
 
 	expectedOutput := map[string]any{
 		"documentdbName": map[string]any{"type": "string"},
 		"location":       map[string]any{"defaultValue": "[resourceGroup().location]", "type": "string"},
-		"mongodbName":    map[string]any{"type": "string"},
 	}
 
 	recipeData, err := driver.GetRecipeMetadata(ctx, BaseOptions{
@@ -515,30 +521,31 @@ func Test_Bicep_GetRecipeMetadata_Success(t *testing.T) {
 }
 
 func Test_Bicep_GetRecipeMetadata_Error(t *testing.T) {
-	t.Skip("This test makes outbound calls. #6490")
+	ts := registrytest.NewFakeRegistryServer(t)
+	t.Cleanup(ts.CloseServer)
+
 	ctx := testcontext.New(t)
-	driver := bicepDriver{}
+	driver := &bicepDriver{RegistryClient: ts.TestServer.Client()}
 	recipeDefinition := recipes.EnvironmentDefinition{
 		Name:         "mongo-azure",
 		Driver:       recipes.TemplateKindBicep,
-		TemplatePath: "ghcr.io/radius-project/dev/test-non-existent-recipe",
+		TemplatePath: ts.TestServer.URL + "/nonexisting:latest",
 		ResourceType: "Applications.Datastores/mongoDatabases",
 	}
 
-	_, err := driver.GetRecipeMetadata(ctx, BaseOptions{
+	_, actualErr := driver.GetRecipeMetadata(ctx, BaseOptions{
 		Recipe:     recipes.ResourceMetadata{},
 		Definition: recipeDefinition,
 	})
 	expErr := recipes.RecipeError{
 		ErrorDetails: v1.ErrorDetails{
 			Code:    recipes.RecipeLanguageFailure,
-			Message: "failed to fetch repository from the path \"ghcr.io/radius-project/dev/test-non-existent-recipe\": ghcr.io/radius-project/dev/test-non-existent-recipe:latest: not found",
+			Message: "failed to fetch repository from the path \"https://<REPLACE_HOST>/nonexisting:latest\": <REPLACE_HOST>/nonexisting:latest: not found",
 		},
 		DeploymentStatus: "setupError",
 	}
-
-	require.Error(t, err)
-	require.Equal(t, err, &expErr)
+	expErr.ErrorDetails.Message = strings.Replace(expErr.ErrorDetails.Message, "<REPLACE_HOST>", ts.URL.Host, -1)
+	require.Equal(t, actualErr, &expErr)
 }
 
 func Test_GetGCOutputResources(t *testing.T) {

--- a/pkg/rp/util/registry.go
+++ b/pkg/rp/util/registry.go
@@ -89,15 +89,6 @@ func getDigestFromManifest(ctx context.Context, repo *remote.Repository, tag str
 	if err != nil {
 		return "", err
 	}
-	layers, ok := manifest["layers"]
-	if !ok {
-		return "", fmt.Errorf("failed to decode the layers from manifest")
-	}
-
-	arr := layers.([]any)
-	if len(arr) == 0 {
-		return "", fmt.Errorf("no layers found in manifest")
-	}
 
 	// get the layers digest to fetch the blob
 	layer, ok := manifest["layers"].([]any)[0].(map[string]any)

--- a/pkg/rp/util/registry.go
+++ b/pkg/rp/util/registry.go
@@ -32,7 +32,7 @@ import (
 // ReadFromRegistry reads data from an OCI compliant registry and stores it in a map. It returns an error if the path is invalid,
 // if the client to the registry fails to be created, if the manifest fails to be fetched, if the bytes fail to be fetched, or if
 // the data fails to be unmarshalled.
-func ReadFromRegistry(ctx context.Context, definition recipes.EnvironmentDefinition, data *map[string]any) error {
+func ReadFromRegistry(ctx context.Context, definition recipes.EnvironmentDefinition, data *map[string]any, client remote.Client) error {
 	registryRepo, tag, err := parsePath(definition.TemplatePath)
 	if err != nil {
 		return v1.NewClientErrInvalidRequest(fmt.Sprintf("invalid path %s", err.Error()))
@@ -42,6 +42,7 @@ func ReadFromRegistry(ctx context.Context, definition recipes.EnvironmentDefinit
 	if err != nil {
 		return fmt.Errorf("failed to create client to registry %s", err.Error())
 	}
+	repo.Client = client
 
 	if definition.PlainHTTP {
 		repo.PlainHTTP = true
@@ -88,6 +89,16 @@ func getDigestFromManifest(ctx context.Context, repo *remote.Repository, tag str
 	if err != nil {
 		return "", err
 	}
+	layers, ok := manifest["layers"]
+	if !ok {
+		return "", fmt.Errorf("failed to decode the layers from manifest")
+	}
+
+	arr := layers.([]any)
+	if len(arr) == 0 {
+		return "", fmt.Errorf("no layers found in manifest")
+	}
+
 	// get the layers digest to fetch the blob
 	layer, ok := manifest["layers"].([]any)[0].(map[string]any)
 	if !ok {

--- a/pkg/rp/util/registrytest/fakeregistryserver.go
+++ b/pkg/rp/util/registrytest/fakeregistryserver.go
@@ -24,6 +24,7 @@ import (
 	"strings"
 	"testing"
 
+	"github.com/go-chi/chi/v5"
 	"github.com/opencontainers/go-digest"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 )
@@ -40,8 +41,13 @@ type fakeServerInfo struct {
 func NewFakeRegistryServer(t *testing.T) fakeServerInfo {
 	blob := []byte(`{
 	"parameters": {
-		"documentdbName": {"type": "string"},
-		"location": {"defaultValue": "[resourceGroup().location]", "type": "string"}
+		"documentdbName": {
+			"type": "string"
+		},
+		"location": {
+			"defaultValue": "[resourceGroup().location]",
+			"type": "string"
+		}
 	}
 }`)
 
@@ -65,35 +71,17 @@ func NewFakeRegistryServer(t *testing.T) fakeServerInfo {
 		Size:      int64(len(index)),
 	}
 
-	prefixURL := "/v2/test"
-
-	ts := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		if r.Method == http.MethodHead {
-			if strings.HasPrefix(r.URL.Path, prefixURL+"/manifests/") {
+	r := chi.NewRouter()
+	r.Route("/v2/test", func(r chi.Router) {
+		r.Route("/manifests", func(r chi.Router) {
+			r.Head("/{ref}", func(w http.ResponseWriter, r *http.Request) {
 				w.Header().Set("Content-Type", indexDesc.MediaType)
 				w.Header().Set("Docker-Content-Digest", indexDesc.Digest.String())
 				w.Header().Set("Content-Length", strconv.Itoa(int(indexDesc.Size)))
 				w.WriteHeader(http.StatusOK)
-			} else if strings.HasPrefix(r.URL.Path, prefixURL+"/blobs/") {
-				w.Header().Set("Content-Type", blobDesc.MediaType)
-				w.Header().Set("Docker-Content-Digest", blobDesc.Digest.String())
-				w.Header().Set("Content-Length", strconv.Itoa(int(blobDesc.Size)))
-				w.WriteHeader(http.StatusOK)
-			} else {
-				w.WriteHeader(http.StatusNotFound)
-			}
+			})
 
-			return
-		} else if r.Method == http.MethodGet {
-			switch r.URL.Path {
-			case prefixURL + "/blobs/" + blobDesc.Digest.String():
-				w.Header().Set("Content-Type", "application/octet-stream")
-				w.Header().Set("Docker-Content-Digest", blobDesc.Digest.String())
-				if _, err := w.Write(blob); err != nil {
-					t.Errorf("failed to write %q: %v", r.URL, err)
-				}
-
-			case prefixURL + "/manifests/" + indexDesc.Digest.String():
+			r.Get("/"+indexDesc.Digest.String(), func(w http.ResponseWriter, r *http.Request) {
 				if accept := r.Header.Get("Accept"); !strings.Contains(accept, indexDesc.MediaType) {
 					t.Errorf("manifest not convertable: %s", accept)
 					w.WriteHeader(http.StatusBadRequest)
@@ -104,18 +92,27 @@ func NewFakeRegistryServer(t *testing.T) fakeServerInfo {
 				if _, err := w.Write(index); err != nil {
 					t.Errorf("failed to write %q: %v", r.URL, err)
 				}
+			})
+		})
 
-			default:
-				t.Errorf("unexpected access: %s %s", r.Method, r.URL)
-				w.WriteHeader(http.StatusNotFound)
-			}
-		} else {
-			t.Errorf("unexpected access: %s %s", r.Method, r.URL)
-			w.WriteHeader(http.StatusMethodNotAllowed)
-			return
-		}
+		r.Route("/blobs", func(r chi.Router) {
+			r.Head("/{digest}", func(w http.ResponseWriter, r *http.Request) {
+				w.Header().Set("Content-Type", blobDesc.MediaType)
+				w.Header().Set("Docker-Content-Digest", blobDesc.Digest.String())
+				w.Header().Set("Content-Length", strconv.Itoa(int(blobDesc.Size)))
+				w.WriteHeader(http.StatusOK)
+			})
+			r.Get("/"+blobDesc.Digest.String(), func(w http.ResponseWriter, r *http.Request) {
+				w.Header().Set("Content-Type", "application/octet-stream")
+				w.Header().Set("Docker-Content-Digest", blobDesc.Digest.String())
+				if _, err := w.Write(blob); err != nil {
+					t.Errorf("failed to write %q: %v", r.URL, err)
+				}
+			})
+		})
+	})
 
-	}))
+	ts := httptest.NewTLSServer(r)
 
 	url, err := url.Parse(ts.URL)
 	if err != nil {

--- a/pkg/rp/util/registrytest/fakeregistryserver.go
+++ b/pkg/rp/util/registrytest/fakeregistryserver.go
@@ -1,0 +1,132 @@
+/*
+Copyright 2023 The Radius Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package registrytest
+
+import (
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strconv"
+	"strings"
+	"testing"
+
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+)
+
+type fakeServerInfo struct {
+	TestServer   *httptest.Server
+	URL          *url.URL
+	CloseServer  func()
+	TestImageURL string
+	ImageName    string
+}
+
+// NewFakeRegistryServer creates a fake registry server that serves a single blob and index.
+func NewFakeRegistryServer(t *testing.T) fakeServerInfo {
+	blob := []byte(`{
+	"parameters": {
+		"documentdbName": {"type": "string"},
+		"location": {"defaultValue": "[resourceGroup().location]", "type": "string"}
+	}
+}`)
+
+	blobDesc := ocispec.Descriptor{
+		MediaType: "recipe",
+		Digest:    digest.FromBytes(blob),
+		Size:      int64(len(blob)),
+	}
+
+	index := []byte(`{
+	"layers": [
+		{
+			"digest": "` + blobDesc.Digest.String() + `"
+		}
+	]
+}`)
+
+	indexDesc := ocispec.Descriptor{
+		MediaType: ocispec.MediaTypeImageIndex,
+		Digest:    digest.FromBytes(index),
+		Size:      int64(len(index)),
+	}
+
+	prefixURL := "/v2/test"
+
+	ts := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Method == http.MethodHead {
+			if strings.HasPrefix(r.URL.Path, prefixURL+"/manifests/") {
+				w.Header().Set("Content-Type", indexDesc.MediaType)
+				w.Header().Set("Docker-Content-Digest", indexDesc.Digest.String())
+				w.Header().Set("Content-Length", strconv.Itoa(int(indexDesc.Size)))
+				w.WriteHeader(http.StatusOK)
+			} else if strings.HasPrefix(r.URL.Path, prefixURL+"/blobs/") {
+				w.Header().Set("Content-Type", blobDesc.MediaType)
+				w.Header().Set("Docker-Content-Digest", blobDesc.Digest.String())
+				w.Header().Set("Content-Length", strconv.Itoa(int(blobDesc.Size)))
+				w.WriteHeader(http.StatusOK)
+			} else {
+				w.WriteHeader(http.StatusNotFound)
+			}
+
+			return
+		} else if r.Method == http.MethodGet {
+			switch r.URL.Path {
+			case prefixURL + "/blobs/" + blobDesc.Digest.String():
+				w.Header().Set("Content-Type", "application/octet-stream")
+				w.Header().Set("Docker-Content-Digest", blobDesc.Digest.String())
+				if _, err := w.Write(blob); err != nil {
+					t.Errorf("failed to write %q: %v", r.URL, err)
+				}
+
+			case prefixURL + "/manifests/" + indexDesc.Digest.String():
+				if accept := r.Header.Get("Accept"); !strings.Contains(accept, indexDesc.MediaType) {
+					t.Errorf("manifest not convertable: %s", accept)
+					w.WriteHeader(http.StatusBadRequest)
+					return
+				}
+				w.Header().Set("Content-Type", indexDesc.MediaType)
+				w.Header().Set("Docker-Content-Digest", indexDesc.Digest.String())
+				if _, err := w.Write(index); err != nil {
+					t.Errorf("failed to write %q: %v", r.URL, err)
+				}
+
+			default:
+				t.Errorf("unexpected access: %s %s", r.Method, r.URL)
+				w.WriteHeader(http.StatusNotFound)
+			}
+		} else {
+			t.Errorf("unexpected access: %s %s", r.Method, r.URL)
+			w.WriteHeader(http.StatusMethodNotAllowed)
+			return
+		}
+
+	}))
+
+	url, err := url.Parse(ts.URL)
+	if err != nil {
+		t.Fatalf("failed to parse url: %v", err)
+	}
+
+	return fakeServerInfo{
+		TestServer:   ts,
+		URL:          url,
+		CloseServer:  ts.Close,
+		TestImageURL: ts.URL + "/test:latest",
+		ImageName:    "test:latest",
+	}
+}

--- a/pkg/rp/util/registrytest/fakeregistryserver.go
+++ b/pkg/rp/util/registrytest/fakeregistryserver.go
@@ -21,7 +21,6 @@ import (
 	"net/http/httptest"
 	"net/url"
 	"strconv"
-	"strings"
 	"testing"
 
 	"github.com/go-chi/chi/v5"
@@ -73,42 +72,34 @@ func NewFakeRegistryServer(t *testing.T) fakeServerInfo {
 
 	r := chi.NewRouter()
 	r.Route("/v2/test", func(r chi.Router) {
-		r.Route("/manifests", func(r chi.Router) {
-			r.Head("/{ref}", func(w http.ResponseWriter, r *http.Request) {
-				w.Header().Set("Content-Type", indexDesc.MediaType)
-				w.Header().Set("Docker-Content-Digest", indexDesc.Digest.String())
-				w.Header().Set("Content-Length", strconv.Itoa(int(indexDesc.Size)))
-				w.WriteHeader(http.StatusOK)
-			})
-
-			r.Get("/"+indexDesc.Digest.String(), func(w http.ResponseWriter, r *http.Request) {
-				if accept := r.Header.Get("Accept"); !strings.Contains(accept, indexDesc.MediaType) {
-					t.Errorf("manifest not convertable: %s", accept)
-					w.WriteHeader(http.StatusBadRequest)
-					return
-				}
-				w.Header().Set("Content-Type", indexDesc.MediaType)
-				w.Header().Set("Docker-Content-Digest", indexDesc.Digest.String())
-				if _, err := w.Write(index); err != nil {
-					t.Errorf("failed to write %q: %v", r.URL, err)
-				}
-			})
+		r.Head("/manifests/{ref}", func(w http.ResponseWriter, r *http.Request) {
+			w.Header().Set("Content-Type", indexDesc.MediaType)
+			w.Header().Set("Docker-Content-Digest", indexDesc.Digest.String())
+			w.Header().Set("Content-Length", strconv.Itoa(int(indexDesc.Size)))
+			w.WriteHeader(http.StatusOK)
 		})
 
-		r.Route("/blobs", func(r chi.Router) {
-			r.Head("/{digest}", func(w http.ResponseWriter, r *http.Request) {
-				w.Header().Set("Content-Type", blobDesc.MediaType)
-				w.Header().Set("Docker-Content-Digest", blobDesc.Digest.String())
-				w.Header().Set("Content-Length", strconv.Itoa(int(blobDesc.Size)))
-				w.WriteHeader(http.StatusOK)
-			})
-			r.Get("/"+blobDesc.Digest.String(), func(w http.ResponseWriter, r *http.Request) {
-				w.Header().Set("Content-Type", "application/octet-stream")
-				w.Header().Set("Docker-Content-Digest", blobDesc.Digest.String())
-				if _, err := w.Write(blob); err != nil {
-					t.Errorf("failed to write %q: %v", r.URL, err)
-				}
-			})
+		r.Get("/manifests/"+indexDesc.Digest.String(), func(w http.ResponseWriter, r *http.Request) {
+			w.Header().Set("Content-Type", indexDesc.MediaType)
+			w.Header().Set("Docker-Content-Digest", indexDesc.Digest.String())
+			if _, err := w.Write(index); err != nil {
+				t.Errorf("failed to write %q: %v", r.URL, err)
+			}
+		})
+
+		r.Head("/blobs/{digest}", func(w http.ResponseWriter, r *http.Request) {
+			w.Header().Set("Content-Type", blobDesc.MediaType)
+			w.Header().Set("Docker-Content-Digest", blobDesc.Digest.String())
+			w.Header().Set("Content-Length", strconv.Itoa(int(blobDesc.Size)))
+			w.WriteHeader(http.StatusOK)
+		})
+
+		r.Get("/blobs/"+blobDesc.Digest.String(), func(w http.ResponseWriter, r *http.Request) {
+			w.Header().Set("Content-Type", "application/octet-stream")
+			w.Header().Set("Docker-Content-Digest", blobDesc.Digest.String())
+			if _, err := w.Write(blob); err != nil {
+				t.Errorf("failed to write %q: %v", r.URL, err)
+			}
 		})
 	})
 

